utils/unittest/mocks/execution_state.go

```diff
@@ -0,0 +1,56 @@
+package mocks
+
+import (
+	"context"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	state "github.com/onflow/flow-go/engine/execution/state/mock"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+// ExecutionState is a mocked version of execution state that
+// simulates some of its behavior for testing purpose
+type ExecutionState struct {
+	sync.Mutex
+	state.ExecutionState
+	commits map[flow.Identifier]flow.StateCommitment
+}
+
+func NewExecutionState(seal *flow.Seal) *ExecutionState {
+	commits := make(map[flow.Identifier]flow.StateCommitment)
+	commits[seal.BlockID] = seal.FinalState
+	return &ExecutionState{
+		commits: commits,
+	}
+}
+
+func (es *ExecutionState) PersistStateCommitment(ctx context.Context, blockID flow.Identifier, commit flow.StateCommitment) error {
+	es.Lock()
+	defer es.Unlock()
+	es.commits[blockID] = commit
+	return nil
+}
+
+func (es *ExecutionState) StateCommitmentByBlockID(ctx context.Context, blockID flow.Identifier) (flow.StateCommitment, error) {
+	commit, ok := es.commits[blockID]
+	if !ok {
+		return nil, storage.ErrNotFound
+	}
+
+	return commit, nil
+}
+
+func (es *ExecutionState) ExecuteBlock(t *testing.T, block *flow.Block) {
+	_, ok := es.commits[block.Header.ParentID]
```

Comment:
I don't think It's necessary, since only concurrent write will need protection. 

There is no write here, only one read from the `commits`.

Also see [this experiment](https://play.golang.org/p/f-zC50dt19r), where it's able to read the value without a lock

Reply:
⚠️ 
The [Go Memory Model](https://golang.org/ref/mem) states:
> Note that a read r may observe the value written by a write w that happens concurrently with r. Even if this occurs, it does not imply that reads happening after r will observe writes that happened before w. 

As far as I understand statement, you are claiming that a read would be able to observe the latest concurrent read even without a lock. This is not true. You will be observing _some_ writes, but there is no guarantee which ones and when without a mutex.